src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/RenameTypeService.cs

```diff
@@ -0,0 +1,239 @@
+                {
+                    if (TryGetCodeElementToRename(oldName, element.Children, isCaseSensitive, out codeElementToRename))
+                    {
+                        return true;
+                    }
+                }
+
+                if (element.Kind == EnvDTE.vsCMElement.vsCMElementStruct ||
+                    element.Kind == EnvDTE.vsCMElement.vsCMElementClass ||
+                    element.Kind == EnvDTE.vsCMElement.vsCMElementModule ||
+                    element.Kind == EnvDTE.vsCMElement.vsCMElementInterface ||
+                    element.Kind == EnvDTE.vsCMElement.vsCMElementEnum)
+                {
+                    string elementName = element.Name;
+                    if(string.Equals(oldName, elementName, isCaseSensitive ? StringComparison.Ordinal: StringComparison.OrdinalIgnoreCase))
+                    {
+                        codeElementToRename = element as EnvDTE80.CodeElement2;
+                        return true;
+                    }
+                }
+            }
+
+            codeElementToRename = null;
+            return false;
+        }
+
+        private static async Task<(bool success, ISymbol symbolToRename)> TryGetSymbolToRenameAsync(string oldName,
+            string newFileName,
+            Project project)
+        {
+            if (project is null)
+                return (false, null);
+
+            Document newDocument = GetDocument(project, newFileName);
+            if (newDocument is null)
+                return (false, null);
+
+            SyntaxNode root = await GetRootNode(newDocument);
+            if (root is null)
+                return (false, null);
+
+            SemanticModel semanticModel = await newDocument.GetSemanticModelAsync();
+            if (semanticModel is null)
+                return (false, null);
+
+            IEnumerable<SyntaxNode> declarations = root.DescendantNodes().Where(n => HasMatchingSyntaxNode(semanticModel, n, oldName, semanticModel.Compilation.IsCaseSensitive));
+            SyntaxNode declaration = declarations.FirstOrDefault();
+            if (declaration is null)
+                return (false, null);
+
+            return (true, semanticModel.GetDeclaredSymbol(declaration));
+
+        }
+
+        private static Document GetDocument(Project project, string filePath)
+            => (from d in project.Documents where StringComparers.Paths.Equals(d.FilePath, filePath) select d).FirstOrDefault();
+
+        private static Task<SyntaxNode> GetRootNode(Document newDocument, CancellationToken token = default) => newDocument.GetSyntaxRootAsync(token);
+
+        private static bool HasMatchingSyntaxNode(SemanticModel model, SyntaxNode syntaxNode, string name, bool isCaseSensitive, CancellationToken token = default)
+        {
+            if (model.GetDeclaredSymbol(syntaxNode, token) is INamedTypeSymbol symbol &&
+                (symbol.TypeKind == TypeKind.Class
```

Comment:
This filter filters to all top-level symbol types you'd already get from GetDeclaredSymbol. The only thing it might be meaningfully filtering out is TypeParameters, and then I'm not sure why you're decending through those in the first place.